### PR TITLE
allow overwriting the output of `rustc --version`

### PR DESCRIPTION
**Pull request by oli-obk**
Our wonderful bisection folk [have to work around](https://github.com/rust-lang/rust/issues/123276#issuecomment-2075001510) crates that do incomplete nightly/feature detection, as otherwise the bisection just points to where the feature detection breaks, and not to the actual breakage they are looking for.

This is also annoying behaviour to nightly users who did not opt-in to those nightly features. Most nightly users want to be in control of the nightly breakage they get, by

* choosing when to update rustc
* choosing when to update dependencies
* choosing which nightly features they are willing to take the breakage for

The reason this breakage occurs is that the build script of some crates run `rustc --version`, and if the version looks like nightly or dev, it will enable nightly features. These nightly features may break in random ways whenever we change something in nightly, so every release of such a crate will only work with a small range of nightly releases. This causes bisection to fail whenever it tries an unsupported nightly, even though that crate is not related to the bisection at all, but is just an unrelated dependency.

This PR (and the policy I want to establish with this FCP) is only for situations like the `version_check`'s `supports_feature` function. It is explicitly not for `autocfg` or similar feature-detection-by-building-rust-code, irrespective of my opinions on it and the similarity of nightly breakage that can occur with such schemes. These cause much less breakage, but should the breakage become an issue, they should get covered by this policy, too.

This PR allows changing the version and release strings reported by `rustc --version` via the `RUSTC_OVERRIDE_VERSION_STRING` env var. The bisection issue is then fixed by https://github.com/rust-lang/cargo-bisect-rustc/pull/335.

I mainly want to establish a compiler team policy: 

> We do not consider feature detection on nightly (on stable via compiler version numbering is fine) a valid use case that we need to support, and if it causes problems, we are at liberty to do what we deem best - either actively working to prevent it or to actively ignore it. We may try to work with responsive and cooperative authors, but are not obligated to.

Should they subvert the workarounds that nightly users or cargo-bisect-rustc can use, we should be able to land rustc PRs that target the specific crates that cause issues for us and outright replace their build script's logic to disable nightly detection.

I am not including links to crates, PRs or issues here, as I don't actually care about the specific use cases and don't want to make it trivial to go there and leave comments. This discussion is going to be interesting enough on its own, without branching out.

### CODE REVIEW

File: compiler/rustc_mir_build/messages.ftl [hunk 1]
```diff
@@ -109,6 +109,9 @@ mir_build_extern_static_requires_unsafe_unsafe_op_in_unsafe_fn_allowed =
     .note = extern statics are not controlled by the Rust type system: invalid data, aliasing violations or data races will cause undefined behavior
     .label = use of extern static
 
+mir_build_feature_detection = Rust language and library feature detection via `version_check::feature_enabled` does not actually check whether the feature works correctly
+    .note = the fucntion has been disabled and will always return `None` on a nightly compiler.
```
Comment (workingjubilee) [inline compiler/rustc_mir_build/messages.ftl:113]:
```suggestion
    .note = the function has been disabled and will always return `None` on a nightly compiler.
```

File: tests/ui/feature-gates/version_check.stderr [hunk 2]
```diff
@@ -0,0 +1,17 @@
+warning: Rust language and library feature detection via `version_check::feature_enabled` does not actually check whether the feature works correctly
+  --> $DIR/version_check.rs:3:1
+   |
+LL | fn supports_feature(_: &str) -> Option<bool> {
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+   = warning: nightly feature detection is disabled
+   = note: for more information, see TODO
+note: the fucntion has been disabled and will always return `None` on a nightly compiler.
```
Comment (workingjubilee) [inline tests/ui/feature-gates/version_check.stderr:9]:
```suggestion
note: the function has been disabled and will always return `None` on a nightly compiler.
```

**Comment by wesleywiser** [comment]
I have a few concerns with this approach:

1. By linting on MIR build of the `version_check::supports_feature`, I believe this means we'll generate this lint anytime the crate is built even if the function isn't used. There are ways to use `version_check` correctly (such as using it to check if rustc's version is a specific stable release number and then enabling the use of language/libstd features stabilized in that release) so incorrectly linting on cases that are written correctly seems very bad. 
    - We could instead lint on calls to `version_check::supports_feature` which would be resolve that particular issue and provide a lint span that is far more actionable than the current experience (most commonly "an indirect dependency depends on `version_check` but we won't tell you which one(s)"). 
2. For crate authors that really want to do this approach, I suspect they will simply copy the relevant parts out of `version_check` directly into their `build.rs` and then remove the dependency, nullifying the lint. As you mention, there are other crates as well that can be problematic when trying to perform bisections. I would rather see a more comprehensive solution that doesn't specifically target one crate in particular. 
    - For instance, we could support a `RUST_ANTI_BOOTSTRAP` environment variable which would cause the compiler to report itself as a stable compiler and ignore the `RUST_BOOTSTRAP` environment variable. Perhaps more usefully, `RUSTC_PRETEND_TO_BE="whatever rustc --version output you want here"` would allow completely overriding the normal version output. `cargo-bisect-rustc` could then be taught to set these flags automatically for the user. Something like this would resolve the issue for a number of similar crates without singling any of them out.

**Comment by oli-obk** [comment]
We already have `RUSTC_FORCE_RUSTC_VERSION`. I have added a test for it and made cargo-bisect-rustc support it with just a flag: https://github.com/rust-lang/cargo-bisect-rustc/pull/335

**Comment by petrochenkov** [comment]
I read the description, but I still don't understand what is `version_check::supports_feature`, how it is related to the compiler, and what this PR does exactly.

**Comment by jackh726** [comment]
I'm a little...lost here. Digging around, if I understand correctly: `ahash` does feature detection to enable `stdsimd` on nightly, which was removed in recent nightlies. And because of that, there are crates that will no longer compile on the latest nightly (but will on the latest stable). In this case, this made finding a regression commit difficult.

The proposed policy in this FCP is to not just *discourage* this use but actively prevent it? That definitely seems a bit too far for me.

I think it definitely makes sense to treat the compilers as "stable" when bisecting - particularly because we don't have any guarantees of stability for nightly features.

I also think it might be worth thinking about why things like `supports_feature` *exists* and work to bring the language and compiler to a point that we can support to cases (and thus also have tight control on how this support changes for things like "pretending to be stable for bisecting").

**Comment by estebank** [comment]
Adding a flag to tell `rustc` to behave as stable as the PR does now will close https://github.com/rust-lang/rust/issues/123404.

---

It should help with a (sufficient?) subset of the problematic cases that the PR initially wanted to address.

There will be cases where the bisection will be for crates using nightly features and a combination of those cases + incorrect rust version checking will continue to manifest the problem of bisecting making crates like `ahash` incorrectly enabling nightly features that will fail at quasi-random places in the bisection.

---

+1 on landing `RUSTC_FORCE_RUSTC_VERSION` (maybe as `RUSTC_FORCE_VERSION`?), after updating the PR title/description.

**Comment by oli-obk** [comment]
I expanded the main PR description.

**Comment by oli-obk** [comment]
> I also think it might be worth thinking about why things like `supports_feature` _exists_ and work to bring the language and compiler to a point that we can support to cases (and thus also have tight control on how this support changes for things like "pretending to be stable for bisecting").

That has the opposite effect of what I'm proposing, as it endorses nightly detection. Nightly detection is absolutely incompatible with semver. It means that users testing their stable crate on nightly to be aware of upcoming changes get random failures from crates that work fine on stable.

**Comment by estebank** [comment]
I appreciate the desire by crate authors to make things "just work™️", but it would be much better for the `build.rs` scripts making feature detection to provide a uses visible message *asking them* to enable feature flags explicitly. This will have a much smaller chance of causing things going awry.

**Comment by chorman0773** [comment]
One thing I'd like is a way to say "Print whatever `$RUST_VERSION` the compiler is built for, but leave off nightly/dev/beta qualifiers". I've run into this with a number of projects, which will be a pain when I package various projects up in tarballs that shouldn't stop building (and shouldn't take on a defacto MxSRV) in the future because someone assumed that X nightly feature will exist forever.

**Comment by chorman0773** [comment]
cargo doesn't swallow stderr IIRC.

**Comment by thomcc** [comment]
Yes it does.

**Comment by oli-obk** [comment]
Yea, even https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargo-warning documents that the messages are hidden for crates.io deps

**Comment by wesleywiser** [comment]
> I mainly want to establish a compiler team policy:
>> We do not consider feature detection on nightly (on stable via compiler version numbering is fine) a valid use case that we need to support, and to actively work to prevent it if it causes problems and the author is not responsive or cooperative.

> Should they subvert the workarounds that nightly users or cargo-bisect-rustc can use, we should be able to land rustc PRs that target the specific crates that cause issues for us and outright replace their build script's logic to disable nightly detection.
> If we as T-compiler establish such a policy, I will also talk with T-crates-io whether this can be considered part of of their usage policy (or become explicitly mentioned in there). That way we have a much faster path to resolving such issues (getting publish access to the relevant crates and disabling their nightly-feature detection logic).

The proposed compiler-team policy makes sense to me especially for crates that actively try to circumvent https://github.com/rust-lang/cargo-bisect-rustc/pull/335. 

I don't think we should ask T-crates-io to adopt the later usage policy. Seizing control of packages that attempt to circumvent 
`RUSTC_FORCE_RUSTC_VERSION` feels like an unreasonably strong action to take and will likely provoke a strong community response. _If_ we're ever in such a situation, we can always take additional technical measures and we could also ban such crates from crater and publish the list of banned crates with rationale on either forge.rlo or even the Rust blog itself. 

@rfcbot concern subsequent-crates-io-policy

**Comment by oli-obk** [comment]
I removed the section on T-crates-io

**Comment by oli-obk** [comment]
Note that banning it from crater means also banning all crates depending on it from crater.

**Comment by wesleywiser** [comment]
@rfcbot resolve subsequent-crates-io-policy

**Comment by oli-obk** [comment]
The text has already been updated, or did I miss something?

**Comment by jackh726** [comment]
> The text has already been updated, or did I miss something?

huh weird, either I'm dumb or it wasn't updated for me. But yeah, you're right.

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
Thanks @oli-obk and others for talking through the design of this, I'm very happy with the consensus reached! 💛 

@bors r+ rollup

**Comment by matthiaskrgr** [comment]
@bors r-
https://github.com/rust-lang/rust/pull/127533#issuecomment-2218685479

**Comment by bjorn3** [comment]
Please use a separate env var for changing the output of `rustc --version`. Setting `RUSTC_FORCE_RUSTC_VERSION` affects symbol names (breaking the protection against mixing rust dylibs compiled by different rustc versions), and breaks the compatibility check for incr comp caches. `RUSTC_FORCE_RUSTC_VERSION` only exists to reduce noise in rustc-perf caused by items getting sorted differently as a result of `DefPathHash`es and symbol names changing based on the rustc version string.

File: tests/ui/feature-gates/version_check.rs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+//@ run-pass
```
Comment (workingjubilee) [inline tests/ui/feature-gates/version_check.rs:1]:
```suggestion
//@ run-pass
//@ ignore-wasm
//@ ignore-emscripten
//@ ignore-sgx
// FIXME: this should be more like //@ needs-subprocesses
```

Comment (jieyouxu) [reply tests/ui/feature-gates/version_check.rs:1]: https://github.com/rust-lang/rust/issues/128295

**Pull request merged**